File: python/42. Trapping Rain Water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        res = 0
        l, r = 0, len(height) - 1
        while l < r:
            minh = min(height[l], height[r])
            if height[l] == minh:
                l += 1
                while l < r and height[l] < minh:
                    res += minh - height[l]
                    l += 1
            else:
                r -= 1
                while l < r and height[r] < minh:
                    res += minh - height[r]
                    r -= 1
        return res

File: go/src/leetcode/42. Trapping Rain Water/42. Trapping Rain Water.go
package main

import "fmt"

func trap(height []int) int {
	ans := 0

	l, r := 0, len(height)-1
	for l < r {
		minHeight := min(height[l], height[r])
		if height[l] == minHeight {
			l++
			for l < r && height[l] < minHeight {
				ans += minHeight - height[l]
				l++
			}
		} else {
			r--
			for l < r && minHeight > height[r] {
				ans += minHeight - height[r]
				r--
			}
		}
	}
	return ans
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(trap([]int{0,1,0,2,1,0,1,3,2,1,2,1}))
}
